### PR TITLE
SNOW-1769605 Add timeout to cloud storage

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description

SNOW-1769605 Added timeout configuration when calling cloud storage

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: azure_storage_client.go [hunk 1]
```diff
@@ -85,9 +86,11 @@ func (util *snowflakeAzureClient) getFileHeader(meta *fileMetadata, filename str
 	if meta.mockAzureClient != nil {
 		blobClient = meta.mockAzureClient
 	}
-	resp, err := blobClient.GetProperties(context.Background(), &blob.GetPropertiesOptions{
-		AccessConditions: &blob.AccessConditions{},
-		CPKInfo:          &blob.CPKInfo{},
+	resp, err := withCloudStorageTimeout(util.cfg, func(ctx context.Context) (blob.GetPropertiesResponse, error) {
+		return blobClient.GetProperties(ctx, &blob.GetPropertiesOptions{
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply azure_storage_client.go:90]: Welcome to the go world :)

File: azure_storage_client.go [hunk 2]
```diff
@@ -279,7 +286,9 @@ func (util *snowflakeAzureClient) nativeDownloadFile(
 		blobClient = meta.mockAzureClient
 	}
 	if meta.options.GetFileToStream {
-		blobDownloadResponse, err := blobClient.DownloadStream(context.Background(), &azblob.DownloadStreamOptions{})
+		blobDownloadResponse, err := withCloudStorageTimeout(util.cfg, func(ctx context.Context) (azblob.DownloadStreamResponse, error) {
+			return blobClient.DownloadStream(context.Background(), &azblob.DownloadStreamOptions{})
```
Comment (sfc-gh-mkubik) [inline azure_storage_client.go:290]: should it be `context` or rather `ctx`? (here and in couple different places)

File: dsn.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ const (
 	defaultRequestTimeout         = 0 * time.Second   // Timeout for retry for request EXCLUDING clientTimeout
 	defaultJWTTimeout             = 60 * time.Second
 	defaultExternalBrowserTimeout = 120 * time.Second // Timeout for external browser login
+	defaultCloudStorageTimeout    = -1                // Timeout for calling cloud storage.
```
Comment (sfc-gh-mkubik) [inline dsn.go:28]: -1 default means that there's no timeout and client is going to wait endlessly?

File: azure_storage_client.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply azure_storage_client.go]: Good catch! I only found this and one another (s3 client) missing.

File: dsn.go (hunk 3 above)
Comment (sfc-gh-pfus) [reply dsn.go:28]: Yes, this is to be backward compatible.

**Review by sfc-gh-mkubik — APPROVED** [review]

**Review by sfc-gh-dheyman — APPROVED** [review]

**Pull request merged**